test/insert.js

```diff
@@ -0,0 +1,62 @@
+'use strict';
+
+var S = require('..');
+
+var eq = require('./internal/eq');
+var jsc = require('jsverify');
+
+
+test('insert', function() {
+
+  eq(typeof S.insert, 'function');
+  eq(S.insert.length, 3);
+  eq(S.insert.toString(), 'insert :: String -> a -> StrMap a -> StrMap a');
+
+  function testBasicInsert(key, val, strMap, strMapResult) {
+    eq(S.insert(key, val, strMap), strMapResult);
+  }
+
+  function unique(xs) {
+    if (xs.length === 0) return [];
+
+    var sorted = S.sort(xs);
+    var x = sorted[0];
+    return sorted.slice(1).reduce(function(acc, x) {
+      return S.equals(x, acc.prev) ? acc : {xs: S.append(x, acc.xs), prev: x};
+    }, {xs: [x], prev: x}).xs;
+  }
+
+  function testKeysOfInsert(key, val, strMap) {
+    eq(
+      unique(S.sort(S.concat(S.keys(strMap), [key]))),
+      S.sort(S.keys(S.insert(key, val, strMap)))
+    );
+  }
```

Comment:
Correct me if I'm wrong, but I don't think this tests anything not covered by other tests is this file. If we were to test `new Point()` the equivalence would not hold because `S.keys(new Point())` is `[]` rather than `['x', 'y']`. This suggests that `S.keys` cannot be implemented in terms of `Object.keys`. This changes the way I feel about sanctuary-js/sanctuary-type-classes#65. I'll comment on that issue shortly.

Reply:
I added the `S.keys` test above because of the conversation between you and @kedashoe earlier. I didn't realize `S.keys(new Point())` would yield `[]`. I understand you're going to decide whether to make `S.keys` get keys of enumerable inherited properties or change `S.concat` to not get enumerable inherited properties. I can wait until you decide or remove `testKeysOfInsert` entirely.

Reply:
We'll update the behaviour of `Z.concat` and `S.concat`. The behaviour of `S.keys` will remain as is.

Reply:
So `Z.concat` and `S.concat` no longer include enumerable inherited properties? And I assume then this PR's `S.insert` and `S.remove` should also no longer include enumerable inherited properties. Then I'll no longer test for enumerable inherited properties or use `new Point()`. Or I could explicitly test that enumerable inherited properties are _not_ included in the results of `S.insert` and `S.remove`?

Do you want me to keep or delete the `testKeysOfInsert` code?

Reply:
@davidchambers just following up on this.

Reply:
Thanks for bringing this pull request to my attention again, Scott. The past few weeks have been busy for me.

I suggest we do the following:

1.  I'll merge the outstanding pull request for sanctuary-type-classes and release v6.1.0.
2.  I'll open a pull request for the change to `Z.concat`.
3.  I'll merge the pull request opened in the previous step.
4.  I'll release v7.0.0.
5.  We'll update to `sanctuary-type-classes@7.0.0`.
6.  You'll rebase this pull request and update the assertions accordingly.
7.  I'll merge this pull request.

We needn't do things in this order, of course, but it would be nice to avoid adding new functions to Sanctuary only to change their behaviour in the following release.

Reply:
Makes sense. I'll be ready for step 6 :). Am I correct in understanding that `Z.concat` and `S.concat` no longer include enumerable inherited properties?

Reply:
> Am I correct in understanding that `Z.concat` and `S.concat` no longer include enumerable inherited properties?

Just about. `Z.concat` and `S.concat` *will* no longer include enumerable inherited properties. ;)

Reply:
sanctuary-js/sanctuary-type-classes#70

Reply:
@samueller, I've completed steps 1–5 in the list above so you should no longer be blocked. :)

Reply:
#439 introduced a StrMap section, so when you rebase please also update the commit message prefix from `object:` to `strmap:`.